In [18]:
## Cleaning for Open AI API
import pandas as pd
import emoji
import re
import string

# Load data from CSV files
test_data = pd.read_csv("./data/issues/issues_test.csv")
train_data = pd.read_csv("./data/issues/issues_train.csv")




In [19]:
test_data

repo           created_at    label  \
0     facebook/react  2023-08-02 02:26:00      bug   
1     facebook/react  2023-07-17 22:43:05      bug   
2     facebook/react  2023-07-13 19:01:47      bug   
3     facebook/react  2023-06-07 17:26:43      bug   
4     facebook/react  2023-05-31 15:17:41      bug   
...              ...                  ...      ...   
1495   opencv/opencv  2022-01-22 11:52:21  feature   
1496   opencv/opencv  2022-01-16 19:27:55  feature   
1497   opencv/opencv  2022-01-14 22:05:58  feature   
1498   opencv/opencv  2022-01-12 09:14:41  feature   
1499   opencv/opencv  2022-01-06 11:01:32  feature   

                                                  title  \
0     Bug: [18.3.0-canary] renderToString hoists som...   
1     [DevTools Bug]: Chrome extension gets disconne...   
2     [DevTools Bug]: Deprecated __REACT_DEVTOOLS_GL...   
3     [DevTools Bug]: React devtools stuck at Loadin...   
4     Bug: Radio button onChange not called in curre...   
...                                                 ...   
1495                               Task: GCC 12 support   
1496  AudioIO: add dnn speech recognition sample on C++   
1497              Use modern OpenVINO package interface   
1498          TiffEncoder write support more depth type   
1499  tiff need check TIFFTAG_SAMPLEFORMAT, should n...   

                                                   body  
0     <!--\r\n  Please provide a clear and concise d...  
1     ### Website or app\r\n\r\nhttps://react.dev/\r...  
2     ### Website or app\n\nN/A\n\n### Repro steps\n...  
3     ### Website or app\n\ncorporate project (priva...  
4     <!--\r\n  Please provide a clear and concise d...  
...                                                 ...  
1495  Support compilation with GCC 12 and fix tests\...  
1496  ### Pull Request Readiness Checklist\r\n\r\nSe...  
1497  * new cmake options: `WITH_OPENVINO`, `OPENCV_...  
1498  **Merge with extra**: https://github.com/openc...  
1499  ### Pull Request Readiness Checklist\r\n\r\nSe...  

[1500 rows x 5 columns]

In [20]:
train_data

repo           created_at    label  \
0     facebook/react  2023-08-26 06:33:37      bug   
1     facebook/react  2023-07-28 05:16:12      bug   
2     facebook/react  2023-07-13 21:58:31      bug   
3     facebook/react  2023-06-14 02:31:20      bug   
4     facebook/react  2023-06-03 11:29:44      bug   
...              ...                  ...      ...   
1495   opencv/opencv  2022-01-24 10:48:13  feature   
1496   opencv/opencv  2022-01-20 12:40:55  feature   
1497   opencv/opencv  2022-01-15 02:39:22  feature   
1498   opencv/opencv  2022-01-14 15:37:53  feature   
1499   opencv/opencv  2022-01-11 16:30:53  feature   

                                                  title  \
0     [DevTools Bug] Cannot add node "1" because a n...   
1     [DevTools Bug]: Devtools extension build faili...   
2     [DevTools Bug]: Deprecated __REACT_DEVTOOLS_GL...   
3     [DevTools Bug] Cannot remove node "0" because ...   
4     [DevTools Bug] Cannot remove node "103" becaus...   
...                                                 ...   
1495                         core: FP denormals support   
1496  feature: submodule or a class scope for export...   
1497                             Reading BigTiff images   
1498                     Add general broadcasting layer   
1499  Adapt remote inference to operate with NV12 blobs   

                                                   body  
0     ### Website or app\n\nPrivate repo cannot give...  
1     ### Website or app\n\nN/A\n\n### Repro steps\n...  
2     ### Website or app\n\nhttps://github.com/open-...  
3     ### Website or app\n\nlocal\n\n### Repro steps...  
4     ### Website or app\n\nlocalhost\n\n### Repro s...  
...                                                 ...  
1495  relates #21046\r\n\r\n- support x86 SSE FTZ+DA...  
1496  All classes are registered in the scope that c...  
1497  **Merge with extra: https://github.com/opencv/...  
1498  Performance details(broadcasting 1x1 to 16x204...  
1499  ### Pull Request Readiness Checklist\r\n\r\nSe...  

[1500 rows x 5 columns]

In [23]:
# Initialize counters for text cleaning
cleaned_count = 0
original_count = 0

# Text cleaning function
def clean_text(text):
    global cleaned_count, original_count

    if not isinstance(text, str):
        original_count += 1
        return text

    # Remove double quotation marks
    text = text.replace('"', '')

    # Remove text starting with "DevTools" and ending with "(automated)"
    text = re.sub(r'DevTools.*?\(automated\)', '', text)

    # Lowercasing should be one of the first steps to ensure uniformity
    text = text.lower()

    # Remove emojis
    text = emoji.demojize(text)

    # Remove URLs
    text = re.sub(r'https?://\S+|www\.\S+', '', text)

    # Remove HTML tags
    text = re.sub(r'<.*?>', '', text)

    # Remove special characters and punctuation
    text = re.sub(f"[{re.escape(string.punctuation)}]", '', text)

    # Remove '#' characters
    text = text.replace("#", "")



    # Remove consecutive whitespaces and replace with a single space
    text = re.sub(r'\s+', ' ', text)

    # Split the text into words
    words = text.split()

    # Remove words that are over 20 characters
    words = [word for word in words if len(word) <= 20]

    # Join the remaining words back into cleaned text
    cleaned_text = ' '.join(words)

    cleaned_count += 1
    return cleaned_text

test_data['body'] = test_data['body'].apply(clean_text)
test_data['title'] = test_data['title'].apply(clean_text)


print(f"Cleaned {cleaned_count} times.")
print(f"Returned original text {original_count} times.")

train_data['body'] = train_data['body'].apply(clean_text)
train_data['title'] = train_data['title'].apply(clean_text)


print(f"Cleaned {cleaned_count} times.")
print(f"Returned original text {original_count} times.")

Cleaned 2998 times.
Returned original text 2 times.
Cleaned 5998 times.
Returned original text 2 times.


In [25]:
test_data

repo           created_at    label  \
0     facebook/react  2023-08-02 02:26:00      bug   
1     facebook/react  2023-07-17 22:43:05      bug   
2     facebook/react  2023-07-13 19:01:47      bug   
3     facebook/react  2023-06-07 17:26:43      bug   
4     facebook/react  2023-05-31 15:17:41      bug   
...              ...                  ...      ...   
1495   opencv/opencv  2022-01-22 11:52:21  feature   
1496   opencv/opencv  2022-01-16 19:27:55  feature   
1497   opencv/opencv  2022-01-14 22:05:58  feature   
1498   opencv/opencv  2022-01-12 09:14:41  feature   
1499   opencv/opencv  2022-01-06 11:01:32  feature   

                                                  title  \
0     bug 1830canary rendertostring hoists some tags...   
1     devtools bug chrome extension gets disconnecte...   
2                               devtools bug deprecated   
3     devtools bug react devtools stuck at loading r...   
4     bug radio button onchange not called in curren...   
...                                                 ...   
1495                                task gcc 12 support   
1496     audioio add dnn speech recognition sample on c   
1497              use modern openvino package interface   
1498          tiffencoder write support more depth type   
1499  tiff need check tifftagsampleformat should not...   

                                                   body  
0     please provide a clear and concise description...  
1     website or app repro steps steps 1 go to a rea...  
2     website or app na repro steps hi i have heard ...  
3     website or app corporate project private repro...  
4     please provide a clear and concise description...  
...                                                 ...  
1495  support compilation with gcc 12 and fix tests ...  
1496  pull request readiness checklist see details a...  
1497  new cmake options withopenvino old options are...  
1498  merge with extra pull request readiness checkl...  
1499  pull request readiness checklist see details a...  

[1500 rows x 5 columns]

In [26]:
train_data

repo           created_at    label  \
0     facebook/react  2023-08-26 06:33:37      bug   
1     facebook/react  2023-07-28 05:16:12      bug   
2     facebook/react  2023-07-13 21:58:31      bug   
3     facebook/react  2023-06-14 02:31:20      bug   
4     facebook/react  2023-06-03 11:29:44      bug   
...              ...                  ...      ...   
1495   opencv/opencv  2022-01-24 10:48:13  feature   
1496   opencv/opencv  2022-01-20 12:40:55  feature   
1497   opencv/opencv  2022-01-15 02:39:22  feature   
1498   opencv/opencv  2022-01-14 15:37:53  feature   
1499   opencv/opencv  2022-01-11 16:30:53  feature   

                                                  title  \
0     devtools bug cannot add node 1 because a node ...   
1     devtools bug devtools extension build failing ...   
2                               devtools bug deprecated   
3     devtools bug cannot remove node 0 because no m...   
4     devtools bug cannot remove node 103 because no...   
...                                                 ...   
1495                          core fp denormals support   
1496  feature submodule or a class scope for exporte...   
1497                             reading bigtiff images   
1498                     add general broadcasting layer   
1499  adapt remote inference to operate with nv12 blobs   

                                                   body  
0     website or app private repo cannot give access...  
1     website or app na repro steps been working on ...  
2     website or app repro steps hi i have heard tha...  
3     website or app local repro steps open reactdev...  
4     website or app localhost repro steps how often...  
...                                                 ...  
1495  relates 21046 support x86 sse ftzdaz flags inf...  
1496  all classes are registered in the scope that c...  
1497  merge with extra resolves 18717 pull request r...  
1498  performance detailsbroadcasting 1x1 to 16x2048...  
1499  pull request readiness checklist see details a...  

[1500 rows x 5 columns]

## Fine Tuning

We are fine tuning chatgpt-3.5-turbo with the training data hopping to accomplish better than results than regularly calling the openai api gpt-4 model

## Data Transformation

Before using the training data for fine tuning, we need to transform our dataframe into a json line file that will be used as every prompt.

The prompt will contain the title and body for every pull request and the answer that we are expecting from our fine tuned model is the respective label for each PR (bug, question, or feature)

In [13]:
import json

# Open the file in write mode
with open('conversational_data.jsonl', 'w', encoding='utf-8') as f:
    # Iterate over the rows in the DataFrame
    for index, row in train_data.iterrows():
        # Create the user message by formatting the prompt with the title and body
        user_message = f"Classify, IN ONLY 1 WORD, the following GitHub pull request as 'feature', 'bug', or 'question' based on its title and body:\n{row['title']}\n{row['body']}"
        
        # Create the assistant message by taking the label
        assistant_message = row['label']
        
        # Construct the conversation object
        conversation_object = {
            "messages": [
                {"role": "system", "content": "GitHub Pull Request Classifier"},
                {"role": "user", "content": user_message},
                {"role": "assistant", "content": assistant_message}
            ]
        }
        
        # Write the conversation object to one line in the file
        f.write(json.dumps(conversation_object, ensure_ascii=False) + '\n')

## Trainig file

Now that we created our json line file that will be used as the initial conversation for our fine tuned model we can upload a trainig file to the OpenAI API

In [31]:
## Uplopading a training file

from openai import OpenAI
client = OpenAI(api_key="open-ai-key")

client.files.create(
  file=open("conversational_data.jsonl", "rb"),
  purpose="fine-tune"
)

FileObject(id='file-wKryp9fKvX2PZUBJ3rw1U3Ag', bytes=2405383, created_at=1699655656, filename='conversational_data.jsonl', object='file', purpose='fine-tune', status='processed', status_details=None)

## Model creation

Finally, we can then create our model. I decided to give the suffix of prissueclassifier.

In [33]:
## Creating a fine-tuned model
client.fine_tuning.jobs.create(
  training_file="file-wKryp9fKvX2PZUBJ3rw1U3Ag", 
  model="gpt-3.5-turbo",
  suffix= "prissueclassifier"
)

FineTuningJob(id='ftjob-BJCtwtaLkzALoBLQAB9Whuu2', created_at=1699655688, error=None, fine_tuned_model=None, finished_at=None, hyperparameters=Hyperparameters(n_epochs='auto', batch_size='auto', learning_rate_multiplier='auto'), model='gpt-3.5-turbo-0613', object='fine_tuning.job', organization_id='org-CpaRU3Zq9ePCCtbhezmcbgrg', result_files=[], status='validating_files', trained_tokens=None, training_file='file-wKryp9fKvX2PZUBJ3rw1U3Ag', validation_file=None)

In [54]:
# Retrieve the state of a fine-tune
client.fine_tuning.jobs.retrieve('ftjob-BJCtwtaLkzALoBLQAB9Whuu2')

FineTuningJob(id='ftjob-BJCtwtaLkzALoBLQAB9Whuu2', created_at=1699655688, error=None, fine_tuned_model='ft:gpt-3.5-turbo-0613:personal:prissueclassifier:8JVGoBJF', finished_at=1699659124, hyperparameters=Hyperparameters(n_epochs=3, batch_size=3, learning_rate_multiplier=2), model='gpt-3.5-turbo-0613', object='fine_tuning.job', organization_id='org-CpaRU3Zq9ePCCtbhezmcbgrg', result_files=['file-Y5wTydUyKC2x2VC54oPMv1rr'], status='succeeded', trained_tokens=1467390, training_file='file-wKryp9fKvX2PZUBJ3rw1U3Ag', validation_file=None)

In [53]:
# List up to 10 events from a fine-tuning job
client.fine_tuning.jobs.list_events(fine_tuning_job_id="ftjob-BJCtwtaLkzALoBLQAB9Whuu2", limit=20)

SyncCursorPage[FineTuningJobEvent](data=[FineTuningJobEvent(id='ftevent-YIqGeIWam8ms4tQEtfMKabUR', created_at=1699659129, level='info', message='The job has successfully completed', object='fine_tuning.job.event', data={}, type='message'), FineTuningJobEvent(id='ftevent-07oALHK8hfUXuCn0fq5fmDVb', created_at=1699659126, level='info', message='New fine-tuned model created: ft:gpt-3.5-turbo-0613:personal:prissueclassifier:8JVGoBJF', object='fine_tuning.job.event', data={}, type='message'), FineTuningJobEvent(id='ftevent-AZXyvI0Dh4aUr2V8NOGxGkId', created_at=1699658907, level='info', message='Step 1401/1500: training loss=2.13', object='fine_tuning.job.event', data={'step': 1401, 'train_loss': 2.129162311553955, 'train_mean_token_accuracy': 0.8888888955116272}, type='metrics'), FineTuningJobEvent(id='ftevent-meuwHJHr814Cf6EEmRia3mDV', created_at=1699658682, level='info', message='Step 1301/1500: training loss=1.68', object='fine_tuning.job.event', data={'step': 1301, 'train_loss': 1.682377

## Fine-tuning results

Our model was sucessfully fine-tuned. Unfortunately, it got 2.1292 training loss.

I only set it for 3 epochs and the total cost the fine tuning was U$11.73 and it used 1,467,390 training tokens.

Check the step_metrics.csv for further details on fine-tuning metrics

## Utilizing Fine-tuned model

First I will test the fine tuned model with the first pull request of our testing dataset, just to see if it works fine

In [63]:
correct_label = test_data.iloc[0]['label'].lower()
description = f"{test_data.iloc[0]['title']} \n {test_data.iloc[0]['body']}"
print(f"Correct PR type: {correct_label}")

prompt = f"Classify, IN ONLY 1 WORD, the following GitHub pull request as 'feature', 'bug', or 'question' based on its title and body:\n{description}"
response = client.chat.completions.create(
    model="ft:gpt-3.5-turbo-0613:personal:prissueclassifier:8JVGoBJF",
    messages=[
            {"role": "system", "content": "GitHub Pull Request Classifier"},
            {"role": "user", "content": prompt}
    ]).choices[0].message.content

# Clean the response to keep only letters (and optionally numbers)
predicted_label = re.sub(r'[^A-Za-z]+', '', response).lower().strip()
print(f"Predicted PR type: {predicted_label}")


Correct PR type: bug
Predicted PR type: bug


In [73]:
import openai
import time
import pandas as pd
import re
import concurrent.futures
from sklearn.metrics import f1_score, precision_score, recall_score, confusion_matrix

# Replace 'open-ai-key' with your actual OpenAI API key
openai.api_key = 'open-ai-key'

# Define the executor once, outside the loop

def query_chatgpt(prompt, temperature=0.0, model="ft:gpt-3.5-turbo-0613:personal:prissueclassifier:8JVGoBJF", max_tokens=1000, max_retries=5):
    """
    Function to query ChatGPT-4 with a given prompt, with retries for timeouts.

    :param prompt: Prompt string to send to ChatGPT-4
    :param model: The model to use, default is ChatGPT-4
    :param max_tokens: Maximum number of tokens to generate
    :param max_retries: Maximum number of retries for timeout
    :return: Response from ChatGPT-4 or None if all retries fail
    """
    attempt = 0
    while attempt < max_retries:
        with concurrent.futures.ThreadPoolExecutor() as executor:
            future = executor.submit(
                openai.chat.completions.create,
                model=model,
                messages=[{"role": "system", "content": "GitHub Pull Request Classifier"}, {"role": "user", "content": prompt}],
                max_tokens=max_tokens,
                temperature=temperature
            )
            try:
                response = future.result(timeout=5)  # 5 seconds timeout
                return response.choices[0].message.content
            except concurrent.futures.TimeoutError:
                print(f"Attempt {attempt + 1}/{max_retries} - Request timed out. Retrying...")
                attempt += 1
                continue
            except Exception as e:
                print(f"Attempt {attempt + 1}/{max_retries} - An error occurred: {e}")
                attempt += 1
                continue
    print("Failed to get a response after several retries.")
    return None
    
labels = ['feature', 'bug', 'question']

y_true = []
y_pred = []

iterations = min(300, len(test_data))

# Now let's loop through the test data and classify the GitHub pull requests
for i in range(iterations):
    correct_label = test_data.iloc[i]['label'].lower()
    description = f"{test_data.iloc[i]['title']} \n {test_data.iloc[i]['body']}"
    print(f"Correct PR type: {correct_label}")
    
    prompt = f"Classify, IN ONLY 1 WORD, the following GitHub pull request as 'feature', 'bug', or 'question' based on its title and body:\n{description}"
    response = query_chatgpt(prompt)
    
    if response is None:
        print("Failed to get a response after several retries. Skipping this item.")
        continue  # Skip this iteration and move to the next one
    
    # Clean the response to keep only letters (and optionally numbers)
    predicted_label = re.sub(r'[^A-Za-z]+', '', response).lower().strip()
    print(f"Predicted PR type: {predicted_label}")
    
    # Append to lists for evaluation
    y_true.append(correct_label)
    y_pred.append(predicted_label)
    time.sleep(6)  # Wait for 3 seconds before retrying

# Calculate weighted average F1-score, precision, and recall
f1 = f1_score(y_true, y_pred, labels=labels, average='weighted')
precision = precision_score(y_true, y_pred, labels=labels, average='weighted')
recall = recall_score(y_true, y_pred, labels=labels, average='weighted')

# Calculate confusion matrix
cm = confusion_matrix(y_true, y_pred, labels=labels)
cm_df = pd.DataFrame(cm, index=labels, columns=labels)

# Calculate TP, FP, FN, TN
results = {}
for i, label in enumerate(labels):
    results[label] = {'TP': cm[i, i]}
    results[label]['FP'] = cm[:, i].sum() - cm[i, i]
    results[label]['FN'] = cm[i, :].sum() - cm[i, i]
    results[label]['TN'] = cm.sum() - (results[label]['TP'] + results[label]['FP'] + results[label]['FN'])

# Print results
for label, metrics in results.items():
    print(f"{label}: {metrics}")

# Save results to CSV
results_df = pd.DataFrame(results).T
results_df['F1-score'] = f1
results_df['Recall'] = recall
results_df['Precision'] = precision


results_df.to_csv('confusion_matrix_all.csv')

print(f"Precision = {precision}")
print(f"Recall = {recall}")
print(f"F1-score = {f1}")

Correct PR type: bug
Predicted PR type: bug
Correct PR type: bug
Predicted PR type: bug
Correct PR type: bug
Predicted PR type: question
Correct PR type: bug
Predicted PR type: bug
Correct PR type: bug
Predicted PR type: bug
Correct PR type: bug
Predicted PR type: bug
Correct PR type: bug
Predicted PR type: bug
Correct PR type: bug
Predicted PR type: bug
Correct PR type: bug
Predicted PR type: bug
Correct PR type: bug
Predicted PR type: bug
Correct PR type: bug
Predicted PR type: bug
Correct PR type: bug
Predicted PR type: bug
Correct PR type: bug
Predicted PR type: bug
Correct PR type: bug
Predicted PR type: bug
Correct PR type: bug
Predicted PR type: bug
Correct PR type: bug
Predicted PR type: bug
Correct PR type: bug
Predicted PR type: bug
Correct PR type: bug
Predicted PR type: bug
Correct PR type: bug
Predicted PR type: bug
Correct PR type: bug
Predicted PR type: bug
Correct PR type: bug
Predicted PR type: bug
Correct PR type: bug
Predicted PR type: bug
Correct PR type: bug
Predic

In [74]:
from sklearn.metrics import classification_report
import pandas as pd

# Create a classification report
report = classification_report(y_true, y_pred, labels=['bug', 'feature', 'question'], target_names=['bug', 'feature', 'question'], zero_division=0, output_dict=True)

# Convert the report to a DataFrame
report_df = pd.DataFrame(report).transpose()

# Print the classification report
print(report_df)

# If you need to access individual metrics, you can do so like this:
precision_bug = report['bug']['precision']
recall_bug = report['bug']['recall']
f1_score_bug = report['bug']['f1-score']

print(f"Precision (bug): {precision_bug}")
print(f"Recall (bug): {recall_bug}")
print(f"F1-score (bug): {f1_score_bug}")

              precision    recall  f1-score     support
bug            0.895238  0.940000  0.917073  100.000000
feature        0.865385  0.900000  0.882353  100.000000
question       0.857143  0.780000  0.816754  100.000000
accuracy       0.873333  0.873333  0.873333    0.873333
macro avg      0.872589  0.873333  0.872060  300.000000
weighted avg   0.872589  0.873333  0.872060  300.000000
Precision (bug): 0.8952380952380953
Recall (bug): 0.94
F1-score (bug): 0.9170731707317075


In [75]:
import openai
import time
import pandas as pd
import re
import concurrent.futures
from sklearn.metrics import f1_score, precision_score, recall_score, confusion_matrix

# Replace 'open-ai-key' with your actual OpenAI API key
openai.api_key = 'open-ai-key'

# Define the executor once, outside the loop

def query_chatgpt(prompt, temperature=0.0, model="ft:gpt-3.5-turbo-0613:personal:prissueclassifier:8JVGoBJF", max_tokens=1000, max_retries=5):
    """
    Function to query ChatGPT-4 with a given prompt, with retries for timeouts.

    :param prompt: Prompt string to send to ChatGPT-4
    :param model: The model to use, default is ChatGPT-4
    :param max_tokens: Maximum number of tokens to generate
    :param max_retries: Maximum number of retries for timeout
    :return: Response from ChatGPT-4 or None if all retries fail
    """
    attempt = 0
    while attempt < max_retries:
        with concurrent.futures.ThreadPoolExecutor() as executor:
            future = executor.submit(
                openai.chat.completions.create,
                model=model,
                messages=[{"role": "system", "content": "GitHub Pull Request Classifier"}, {"role": "user", "content": prompt}],
                max_tokens=max_tokens,
                temperature=temperature
            )
            try:
                response = future.result(timeout=5)  # 5 seconds timeout
                return response.choices[0].message.content
            except concurrent.futures.TimeoutError:
                print(f"Attempt {attempt + 1}/{max_retries} - Request timed out. Retrying...")
                attempt += 1
                continue
            except Exception as e:
                print(f"Attempt {attempt + 1}/{max_retries} - An error occurred: {e}")
                attempt += 1
                continue
    print("Failed to get a response after several retries.")
    return None
    
labels = ['feature', 'bug', 'question']

y_true = []
y_pred = []

iterations = min(1500, len(test_data))

# Now let's loop through the test data and classify the GitHub pull requests
for i in range(iterations):
    correct_label = test_data.iloc[i]['label'].lower()
    description = f"{test_data.iloc[i]['title']} \n {test_data.iloc[i]['body']}"
    print(f"Correct PR type: {correct_label}")
    
    prompt = f"Classify, IN ONLY 1 WORD, the following GitHub pull request as 'feature', 'bug', or 'question' based on its title and body:\n{description}"
    response = query_chatgpt(prompt)
    
    if response is None:
        print("Failed to get a response after several retries. Skipping this item.")
        continue  # Skip this iteration and move to the next one
    
    # Clean the response to keep only letters (and optionally numbers)
    predicted_label = re.sub(r'[^A-Za-z]+', '', response).lower().strip()
    print(f"Predicted PR type: {predicted_label}")
    
    # Append to lists for evaluation
    y_true.append(correct_label)
    y_pred.append(predicted_label)
    time.sleep(6)  # Wait for 3 seconds before retrying

# Calculate weighted average F1-score, precision, and recall
f1 = f1_score(y_true, y_pred, labels=labels, average='weighted')
precision = precision_score(y_true, y_pred, labels=labels, average='weighted')
recall = recall_score(y_true, y_pred, labels=labels, average='weighted')

# Calculate confusion matrix
cm = confusion_matrix(y_true, y_pred, labels=labels)
cm_df = pd.DataFrame(cm, index=labels, columns=labels)

# Calculate TP, FP, FN, TN
results = {}
for i, label in enumerate(labels):
    results[label] = {'TP': cm[i, i]}
    results[label]['FP'] = cm[:, i].sum() - cm[i, i]
    results[label]['FN'] = cm[i, :].sum() - cm[i, i]
    results[label]['TN'] = cm.sum() - (results[label]['TP'] + results[label]['FP'] + results[label]['FN'])

# Print results
for label, metrics in results.items():
    print(f"{label}: {metrics}")

# Save results to CSV
results_df = pd.DataFrame(results).T
results_df['F1-score'] = f1
results_df['Recall'] = recall
results_df['Precision'] = precision


results_df.to_csv('confusion_matrix_full.csv')

print(f"Precision = {precision}")
print(f"Recall = {recall}")
print(f"F1-score = {f1}")

Correct PR type: bug
Predicted PR type: bug
Correct PR type: bug
Predicted PR type: bug
Correct PR type: bug
Predicted PR type: question
Correct PR type: bug
Predicted PR type: bug
Correct PR type: bug
Predicted PR type: bug
Correct PR type: bug
Predicted PR type: bug
Correct PR type: bug
Predicted PR type: bug
Correct PR type: bug
Predicted PR type: bug
Correct PR type: bug
Predicted PR type: bug
Correct PR type: bug
Predicted PR type: bug
Correct PR type: bug
Predicted PR type: bug
Correct PR type: bug
Predicted PR type: bug
Correct PR type: bug
Predicted PR type: bug
Correct PR type: bug
Predicted PR type: bug
Correct PR type: bug
Predicted PR type: bug
Correct PR type: bug
Predicted PR type: bug
Correct PR type: bug
Predicted PR type: bug
Correct PR type: bug
Predicted PR type: bug
Correct PR type: bug
Predicted PR type: bug
Correct PR type: bug
Predicted PR type: bug
Correct PR type: bug
Predicted PR type: bug
Correct PR type: bug
Predicted PR type: bug
Correct PR type: bug
Predic

In [77]:
from sklearn.metrics import classification_report
import pandas as pd

# Create a classification report
report = classification_report(y_true, y_pred, labels=['bug', 'feature', 'question'], target_names=['bug', 'feature', 'question'], zero_division=0, output_dict=True)

# Convert the report to a DataFrame
report_df = pd.DataFrame(report).transpose()

# Print the classification report
print(report_df)

# If you need to access individual metrics, you can do so like this:
precision_bug = report['bug']['precision']
recall_bug = report['bug']['recall']
f1_score_bug = report['bug']['f1-score']


              precision    recall  f1-score      support
bug            0.869748  0.839757  0.854489   493.000000
feature        0.862000  0.865462  0.863727   498.000000
question       0.799607  0.823887  0.811565   494.000000
accuracy       0.843098  0.843098  0.843098     0.843098
macro avg      0.843785  0.843035  0.843261  1485.000000
weighted avg   0.843817  0.843098  0.843308  1485.000000
